src/audio/tdfb/tdfb_direction.c

```diff
@@ -0,0 +1,559 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+// Author: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
+
+#include <ipc/topology.h>
+#include <sof/audio/tdfb/tdfb_comp.h>
+#include <sof/lib/alloc.h>
+#include <sof/math/iir_df2t.h>
+#include <sof/math/trig.h>
+#include <sof/math/sqrt.h>
+#include <user/eq.h>
+#include <stdint.h>
+
+/* Generic definitions */
+#define COEF_DEG_TO_RAD		Q_CONVERT_FLOAT(0.017453, 15)	/* Q1.15 */
+#define COEF_RAD_TO_DEG		Q_CONVERT_FLOAT(57.296, 9)	/* Q6.9 */
+#define PI_Q12			Q_CONVERT_FLOAT(3.1416, 12)
+#define PIMUL2_Q12		Q_CONVERT_FLOAT(6.2832, 12)
+#define PIDIV2_Q12		Q_CONVERT_FLOAT(1.5708, 12)
+
+/* Sound levels filtering related, these form a primitive voice activity
+ * detector. Sound levels below ambient estimate times threshold (kind of dB offset)
+ * are not scanned for sound direction.
+ */
+#define SLOW_LEVEL_SHIFT	12
+#define FAST_LEVEL_SHIFT	1
+#define POWER_THRESHOLD		Q_CONVERT_FLOAT(15.849, 10)	/* 12 dB, value is 10^(dB/10) */
+
+/* Iteration parameters, smaller step and larger iterations is more accurate but
+ * consumes more cycles.
+ */
+#define AZ_STEP			Q_CONVERT_FLOAT(0.6, 12)	/* Radians as Q4.12 (~34 deg) */
+#define AZ_ITERATIONS		8				/* loops in min err search */
+#define SOURCE_DISTANCE		Q_CONVERT_FLOAT(3.0, 12)	/* source distance in m Q4.12 */
+
+/* Sound direction angle filtering */
+#define SLOW_AZ_C1		Q_CONVERT_FLOAT(0.02, 15)
+#define SLOW_AZ_C2		Q_CONVERT_FLOAT(0.98, 15)
+
+/* Threshold for notifying user space, no more often than every 100 ms */
+#define CONTROL_UPDATE_MIN_TIME	Q_CONVERT_FLOAT(100e-3, 16)
+
+/* Emphasis filters for sound direction (IIR) */
+
+uint32_t iir_emphasis_48k[20] = {
+	0x00000002, 0x00000002, 0x00000000, 0x00000000,
```

Comment:
These were generated with https://github.com/thesofproject/sof/blob/main/tools/tune/tdfb/example_direction_emphasis.m. I could place these as header file to an include directory for filter coefficients like for SRC. Would that be better?

I didn't make these passed via blob because the emphasis filter shape is for human voice generally. If desire to track something else like ultrasound, these should change.

Reply:
Just a link to the script in a comment is fine i think. Requiring to build tools sounds like too much of a pain. 

And by define directly i meant more define the struct directly rather than casting from an array so it is easy to read params, also valid alternative.

Reply:
Thanks, I added comment text. C struct output option could be interesting for IIR EQ tool (later). By-hand formatting would be very risky for errors. I wonder if the flexible arrays in the end for biquads can be output that way.